File: consumerNodejs/src/rabbitMQ/rabbitmq.service.ts
import { connect, Connection, Channel } from 'amqplib';
import {createId} from "@paralleldrive/cuid2"

export interface MessageTemplateProps<T=any> {
  id: string;
  type: string;
  date: Date;
  payload: T;
}

export class RabbitMQService {
  private connection: Connection;
  private channel: Channel;

  constructor(private readonly prefetch: number = 1) {}

  async connect() {
    this.connection = await connect(process.env.RABBIT_MQ_URL as string);
    this.channel = await this.connection.createChannel();
    console.log("Rabbit connected!")
  }

  async consumer(queueName: string, consumerFunc: (msg:MessageTemplateProps) => Promise<void>)  {
    this.channel.assertQueue(queueName, { durable: true })
    this.channel.prefetch(this.prefetch)
    this.channel.consume(queueName, async (msg)=>{
      if(!msg) return
      const message = JSON.parse(msg.content.toString()) as MessageTemplateProps
      try{
        await consumerFunc(message)
        this.channel.ack(msg)
      }catch(e){
        console.error("ERR_TO_CONSUME_MESSAGE")
        this.channel.nack(msg, undefined, true)
      }

    },{noAck: false})

    console.log(`consuming ${queueName}`)

  }

  async close() {
    await this.channel.close();
    await this.connection.close();
  }
}

File: consumerRust/src/rabbitmq/rabbit_mq_service.rs
use std::{ env, error::Error};
use futures_util::stream::StreamExt;

use lapin::{options::{BasicAckOptions, BasicConsumeOptions, BasicNackOptions, QueueDeclareOptions}, types::FieldTable, ConnectionProperties};

use crate::interfaces::MessageTemplateProps;

use super::MessageConsumerService;

pub struct RabbitMqService {
    channel: lapin::Channel,
    message_consumer_service: MessageConsumerService
}

impl RabbitMqService {
    pub async fn new(message_consumer: MessageConsumerService) -> Result<RabbitMqService, Box<dyn Error>> {
        
        let rabbit_url = env::var("RABBIT_MQ_URL").unwrap();

        let con = lapin::Connection::connect(&rabbit_url, ConnectionProperties::default()).await?;
        let ch = con.create_channel().await?;

        let rabbit = RabbitMqService {
            channel: ch,
            message_consumer_service: message_consumer
        };

        Ok(rabbit)
    }

    pub async  fn consumer(&mut self, queue_name: &str) -> Result<(),Box<dyn Error>> {
        let options = QueueDeclareOptions {
            auto_delete: false,
            durable: true,
            exclusive: false,
            ..Default::default()
        };

        let options_consumer = BasicConsumeOptions{
            exclusive: false,
            no_ack: false,
            no_local: false,
            nowait: false
        };

        let queue = self.channel
        .queue_declare(queue_name, options, FieldTable::default())
        .await?;
    
        let mut consumer = self.channel
            .basic_consume(
                queue.name().as_str(),
                "my_consumer",
                options_consumer,
                FieldTable::default(),
            )
            .await?;
        
        while let Some(delivery) = consumer.next().await {
            if let Ok(delivery) = delivery {
                    let body = String::from_utf8_lossy(&delivery.data);
                   
                    let new_body = body.replace("\"type\"", "\"type_message\"");
                    
                    let message_to_save: MessageTemplateProps = serde_json::from_str(&new_body)?;
                    let res = self.message_consumer_service.consume(message_to_save).await;

                    if let Err(err) = res {
                        delivery.nack(BasicNackOptions{
                            multiple: false,
                            requeue: true
                        }).await?;
                        println!("{}", err.to_string());
                    }

                    delivery.ack(BasicAckOptions::default()).await?;
            };

        }

        /*for (i, message) in consumer.receiver().iter().enumerate() {
            match message {
                ConsumerMessage::Delivery(delivery) => {
                    let body = String::from_utf8_lossy(&delivery.body);
                    println!("({:>3}) Received [{}]", i, body);
                   
                    let new_body = body.replace("\"type\"", "\"type_message\"");
                    
                    let message_to_save: MessageTemplateProps = serde_json::from_str(&new_body)?;
                    let res = self.message_consumer_service.consume(message_to_save).await;

                    if let Err(err) = res {
                        consumer.nack(delivery.clone(), true)?;
                        println!("{}", err.to_string());
                    }

                    consumer.ack(delivery)?;
                }
                other => {
                    println!("Consumer ended: {:?}", other);
                    break;
                }
            }
        };*/
    
        Ok(())

    }
}